clparser.1
.TH clparser 1 "29 May, 2023"
.
.SH NAME
clparser - command line argument parser for shells
.
.SH SYNOPSIS
.\" .SY clparser
.\" .RI [ OPTIONS ]
.\" .B --
.\" .I COMMAND_LINE_ARGS
.\" [\&\.\|.\|.\&]
.\" .YS
.\" .
.SY clparser
.\" .OP \-A|\-\-override-argv
.RB { \-A | \-\-override-argv }
.\" .OP \-a|\-\-maintain-argv
.RB { \-a | \-\-maintain-argv }
.\" .OP \-S|\-s|\-\-shell shell
.RB { \-S | \-s | \-\-shell
.IR shell }
.RB { \-H | \-m | \-\-help\-msg
.IR helpmessage }
.B \-\-
.RI [ COMMAND_LINE_ARGS
\&.\|.\|.\&]
.YS
.
.SY clparser
.BR \-\-help | \-h
.YS
.
.SH DESCRIPTION
.B clparser
takes in command line arguments to parse as command line arguments and a specification for which command line arguments should exist from stdin, and outputs shell code to set variables set from the command line arguments to stdout.
Command line arguments to be parsed should be given after
.BR -- .
This will prevent the command line arguments to be parsed from being interpreted as
.B clparser
command line arguments.
.
.SS Argument Specification Format
The format specifies
.I flags
for setting some optional value to true and
.I parameters
for setting some (optional) value to the value passed in the command line arguments.
There are also
.I positionals
for setting some parameter values (like
.IR parameters )
through named positional arguments rather than specifying the name through
.B \-n
or
.BR \-\-name .
Arguments as specified as flags, parameters, or positionals by being listed after
.IR flags: ,
.IR parameters: ,
or
.IR positionals: ,
respectively.
Arguments are defined as in a space separated list, ended with a semicolon
.RI ( ; ).
E.g.:
.IR "flags: f help test;" .
Arguments to be set together are listed together in a comma separated list.
E.g.:
.IR "flags: f,flags help,h test,t,T;" .
.I positionals
specify the positional arguments in the order given after the
.I positionals:
keyword.
.PP
Parameters (and positionals) can have default values.
Default values set the value of the parameters if the parameter is not set with the input command line arguments.
Default values are set by appending an equals
.RI ( = )
after the parameters (e.g.:
.IR line,l=default ).
.PP
Flags can specify other flags that would negate its value if both are defined.
When a flag with a negation is input and its negation flag is input after it, then the output of the first flag will be false because the second value (set to true) negated it.
E.g.: if
.B \-i
has
.B \-o
set as a negation flag, then the following input arguments
.B \-i \-\-param
.I value
.B \-o
will set set
.B i
to false and
.B o
to true.
This can be helpful, especially when users alias a script with arguments that the user might want to override (imagine
.B \-i
was set in the alias and
.B \-\-param
.I value
.B \-o
was set by the user).
Negation flags can be set by appending equals minus
.RB ( =- )
after flags (e.g.:
.BR i,in=-out ).
Multiple negation flags can be given (e.g.:
.BR i,in=-out=-h ),
which would reset the value of
.B i
if
.B \-\-out
or
.B \-h
are specified after
.BR \-i .
.PP
For a full example, see the
.B EXAMPLES
section below.
.
.SS Parsing Features
There are 2 different types of command line arguments: words and single letters.
Words are specified after 2 dashes
.RB ( -- ),
letters are specified after 1 dash
.RB ( - ).
Letters can be combined on the same input (e.g.:
.BR \-aBc ).
Word parameters can have dashes
.RB ( - )
in them.
These words will be output with underscores
.RB ( _ )
instead of dashes to make a valid variable name.
.PP
Single letter parameters can be used in the combined format if they are the last letter and the value of the parameter is given as the next argument (e.g.:
.BR "\-aBc value" ),
where
.B c
is a parameter, and
.B a
and
.B B
are flags.
.PP
Word parameters can have the value specified in the same argument after an equals
.RB ( = )
(e.g.:
.B \-\-word=\c
.IR value ).
.PP
Positional arguments are defined as values that don't start with \- and do not follow parameters.
If a
.B --
is found, all arguments after it are interpreted as positional arguments.
.SS Description of Output
.B clparser
outputs shell syntax to set variables to the values of the parameters.
By default, the shell will try to choose the shell syntax from the name of the calling process (found in
.RI /proc/ ppid /comm).
If the shell has associative arrays (maps, hash tables, dicts, etc.), 
.I flags
are set in an associative array called
.BR flags ,
and
.I parameters
are set in an associative array called
.BR params .
`sh`, `csh`, and `fish` do not have associative arrays, so flags and parameters are just set as variables.
Positional arguments are set in an array.
`sh` does not have arrays, so it overrides argv to place the positional arguments by default.
See the
.B OPTIONS
section below to see how to override these.
.PP
The output of parameter values and positional arguments is done through the existing positional arguments
.RB ( \(dq${\c
.I number\c
.BR }" )
to prevent issues with the shell interpreting whitespace or symbols in the parameters (see the
.B EXAMPLES
below for an example).
.
.SH OPTIONS
.BR \-h ,
.B \-\-help
Print help info to stderr.
Starts with options showing flags and params in usage for.
Parameters also include the value for the user to input based on the name of the parameter.
If parameters have default values, it shows it after an =.
Under the usage line, it shows lines for each argument with a specified help message (see \-\-help\-msg below).
.PP
.BR \-A ,
.B \-\-override\-argv
Output positional arguments in place of the shell's argv.
This may be useful in some cases, but most shells have arrays to put the arguments in.
This is used by default for `sh` since "$@" is the only way of splitting arguments by argument while maintaining whitespace.
.PP
.BR \-a ,
.B \-\-maintain\-argv
Output positional arguments in a variable named
.B args
instead of overriding argv.
This will allow the original argv to be analyzed and used after parsing arguments.
This will use an array for all shells other than `sh`, since it does not have arrays.
This is not a good option for `sh` since "$@" is the only way of splitting arguments by argument while maintaining whitespace.
.PP
.BR \-n ,
.B \-\-namespace
Namespace args by putting them in associative arrays when available or prepending flags_ or params_ to the variable names.
This option is on by default for shells with associative arrays.
.PP
.BR \-N ,
.B \-\-no\-namespace
Do not namespace args: set variable names directly, allowing flags and parameters to overlap.
This option is on by default for shells without associative arrays.
.PP
.BR \-E ,
.B \-\-no\-output\-empty
Do not output parameters that do not have a value (not specified on command line and no default value).
This can be helpful for a flag and parameter of the same name.
E.g.: a choose flag to indicate to use some user input, but also have a parameter of the same name that takes the user input.
This is only useful for shells that won't error on using undeclared variables that don't use associative arrays (sh and fish) (and namespace isn't used) (or no-namespace is used).
.PP
.BR \-u ,
.BR \-P ,
.B \-\-unknown\-positional
Rather than error on unknown arguments, assume all further arguments are positional.
This is basically allowing users to not need to pass in
.B --
to allow the rest of the arguments to be positional.
.PP
.BR \-e ,
.B \-\-help\-exits
The help option will automatically exit.
This is done by printing `exit` to stdout to be evaluated.
.PP
.BR \-s ,
.BR \-S ,
.B \-\-shell
.I shell
Specify the shell language syntax to be used.
By default, the shell will try to choose the shell syntax from the name of the calling process.
.PP
.BR \-p ,
.B \-\-prog\-name
.I prog_name
Specify the name of the script to use in the usage message.
By default, the shell will try to choose the path of the script of the calling process (the first argument in the cmdline).
This is taken from
.RI /proc/ ppid /cmdline.
.PP
.BR \-H ,
.BR \-m ,
.B \-\-help\-msg
.I helpmsgs
Specify help messages for specified flags/parameters.
The format is
.BR "argument = help message" ,
with argument, message pair per line.
See the
.B EXAMPLES
section below for an example format and output.
.
.SH EXIT STATUS
Exit with success (0) if the arguments and specification are parsed successfully.
Exit with code 1 if outputting help.
Exit with code 1 if there are not alphanumeric (or _ or \-) characters in input parameters.
Exit with code 2 if the passed argument was not found in the specification.
Exit with code 3 if parsing the specification string failed.
Exit with code 2 if there were neither flags or parameters defined in the specification string.
.PP
Yes, these should be separated better.
.
.SH EXAMPLES
Run
.B clparser
in POSIX compliant shells using
.EX sh
spec='flags: f,flag=-g=-h g=-h=-qwerty qwerty=-flag=-h=-g h,help; parameters: q,asdf u=defval nothing zzz,z,Z=someth;'
helpmsg='
help = print this help message
flag = a random flag
g    = i dunno, its nonsense
qwerty = a keyboard layout
asdf = q param
u    = something
zzz  = a bunch of zs
\(aq
eval "$(echo "$spec" | clparser -- "$@")"
.EE
This will output code that looks like
.EX sh
flag=false
f=false
h=false
help=false
g=false
qwerty=true
asdf="${2}"
q="${2}"
Z='test'
zzz='test'
z='test'
u='defval'
nothing="${8}"
set -- "${4}" "${6}" "${10}" "${11}"
.EE
if the user called the script with the arguments
.EX
-fgq whatever --qwerty 'i dunno' --zzz=test words --nothing else -- --test hi
.EE
.PP
For `csh`, call
.B clparser
in this format:
.EX csh
eval `echo $spec:q | ./clparser --help-msg $helpmsg:q -- $argv:q`
.EE
Note that just using
.B $argv
will not preserve whitespace, which will cause bugs, so appending `:q` at the end of variable names is necessary.
.B clparser
and evaluate the resulting `csh` code.
.PP
See `fish` and `xonsh` example syntax in the github repo example files.
For `xonsh`, don't bother using this; just use
.BR argparse .
.PP
The
.B \-\-help
output will look like
.EX
Usage:	"${0}" [ --flag -f ] [ -h --help ] [ -g ] [ --qwerty ] --asdf -q asdf [ -Z --zzz -z zzz = someth ] [ -u u = defval ] --nothing nothing
 [ -h --help ]                                    print this help message
 [ --flag -f ]                                    a random flag
 [ -g ]                                           i dunno, its nonsense
 [ --qwerty ]                                     a keyboard layout
 --asdf -q asdf                                   q param
 [ -u u = defval ]                                something
 [ -Z --zzz -z zzz = someth ]                     a bunch of zs
.EE
.PP
.B TODO
add examples of output for `bash` and `csh` and show use for `fish`.


clparser.c
#define _GNU_SOURCE
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include "parseargs.h"
#include "printargs.h"

#define MAX_SHELL_LEN   10
#define MAX_PROCCOMM_LEN    (6 + 10 + 5 + 1)    // /proc/ + ppid + /comm    + \000
#define MAX_PROCCMDLINE_LEN (6 + 10 + 8 + 1)    // /proc/ + ppid + /cmdline + \000




// TODO add multiple args support like nargs=+ from python argparse
// TODO update return values to be unique (maybe overlap with map errors)

// TODO go through this list of things and correct the things that are wrong and add features and things that i want
// so the idea is that youll send in configuration stuffs over stdin and then have it take in the clargs as clargs and then itll output things that i guess are helpful based on the config stuffs sent in over stdin
// theres no way that this will be overcomplicated or impossible to use or write at all
// ill define that this can have flags and options that are set to values and a default field that everything else falls into
// ill do it in the gnuish i think style where there can be words that use the --, single letters use - and can be combined 
// if the single letter - options arent flags then the next arguments will be used as the options for that parameter thingy i dont know howw to explain
// it should go in order of the options it gets so if they contradict then itll be the last option
// the word parameters can use a space or an = with no spaces
// the stdin should be of the form:
// spec="flags: f,flag=-g=-h g=-h=-qwerty qwerty=-flag=-h=-g h,help; parameters: q,asdf u=defval nothing zzz,z,Z=someth;"
// set default values for args using `a,arg=default_val`
// set other flags to negate the current flag using `f,flag=-other-flag`
//  - this way if other-flag is set after flag, then flag will be reset, no logic needed outside
//  - additionally this means that the last switch is always the most important (useful for overriding aliases)
// where things that are in the same word separated by commas are the same flag or parameter (if its only 1 char its a single char - else its a -- word)
// things after flags are flags and the stuff after parameters are the options with values to be set
// this should return a string like:
//  flag=false
//  f=false
//  h=false
//  help=false
//  g=false
//  qwerty=true
//  asdf="${2}"
//  q="${2}"
//  Z='test'
//  zzz='test'
//  z='test'
//  u='defval'
//  nothing="${8}"
//  set -- "${4}" "${6}" "${10}" "${11}"
// if called with arguments `-fgq whatever --qwerty 'i dunno' --zzz=test words --nothing else -- --test hi`
// run it like eval "$(echo "$spec" | clparser -- "$@")" and itll set the variables accordingly so you can use them without doing any parsing
// error codes:
// 1 not alphanumeric flags or parameter
// 2 flags and parameter not defined
// 3 bad definitions
// use structure echo "$spec" | clparser --args-for-parser -- "$@"
//  in this case the arguments after -- are what needs to be parsed which makes the -- necessary (could make it something like -p or --parse instead of just --)
//  (could have it assume to parse everything if there is no -- but that could be confusing if someone passes in a -- that wasnt exected and things are parsed wrong or it errors cause thats how that works)
// if it hits a -- with no word after it should take that to mean everything after is default since that seems to be how most things would use it (this is less flexible in some ways but i dont want to deal with it)
// ive decided this needs a help system so probably that would be like giving one of the names and then the help for it in a big string and then passing that string into the parser with a like --helpmsgs
// this thing needs a help and other arguments in general probably
int main(int argc, const char * const argv[]){

    // setup internal flags and params maps
    map_t flagMap_loc, paramMap_loc;
    const char ** passedArgs = NULL;
    int passedArgc;
    initMap(&flagMap_loc);
    initMap(&paramMap_loc);
    // TODO add better support for positional parameters:
    //  1.  flag to specify "stop parsing" (count rest of args as positional) if any positional arg found
    //  2.  flag to specify "stop parsing" (count rest of args as positional) if any arg found that is not in the specification
    //      - could encapsulate 1. if you explicitly specify 0 positional args (probably better this way)
    //  3.  add something specific for positional parameters to be able to look up values
    //      - eg: script file1 file2 [options] setting file1 to first arg and file2 to setcond arg rather than just an array
    //      - need some way of specifying which position the arg should be filled by
    //      - need to support both named positional args and existing array style
    //      - maybe have parseArgs take in an array (NULL terminated) of MapData pointers of nodes of the positional args map, in the order of positional args
    //          - get map nodes from return of addMapMembers, just add to array at end
    //          - spec parser just parse like parameters but then create array (TODO should i make a new function or work around the assumptions of the existing one)
    MapData * helpNode         = addMapMembers(& flagMap_loc, &flagFalse, BOOL, false, "Ssd"  , STRVIEW("help"              ), "h", 1);
    MapData * maintainArgvNode = addMapMembers(& flagMap_loc, &flagFalse, BOOL, false, "Ssd"  , STRVIEW("maintain-argv"     ), "a", 1);
    MapData * overrideArgvNode = addMapMembers(& flagMap_loc, &flagFalse, BOOL, false, "Ssd"  , STRVIEW("override-argv"     ), "A", 1);
    MapData * helpExitsNode    = addMapMembers(& flagMap_loc, &flagFalse, BOOL, false, "Ssd"  , STRVIEW("help-exits"        ), "e", 1);
    MapData * useNamespaceNode = addMapMembers(& flagMap_loc, &flagFalse, BOOL, false, "Ssd"  , STRVIEW("namespace"         ), "n", 1);
    MapData *  noNamespaceNode = addMapMembers(& flagMap_loc, &flagFalse, BOOL, false, "Ssd"  , STRVIEW("no-namespace"      ), "N", 1);
    MapData * noOutEmptyNode   = addMapMembers(& flagMap_loc, &flagFalse, BOOL, false, "Ssd"  , STRVIEW("no-output-empty"   ), "E", 1);
    MapData * unkIsPosNode     = addMapMembers(& flagMap_loc, &flagFalse, BOOL, false, "Ssdsd", STRVIEW("unknown-positional"), "u", 1, "P", 1);
    MapData * helpMsgNode      = addMapMembers(&paramMap_loc, NULL      , STR , false, "Ssdsd", STRVIEW("help-msg"          ), "H", 1, "m", 1);
    MapData * shellNode        = addMapMembers(&paramMap_loc, "default" , STR , true , "Ssdsd", STRVIEW("shell"             ), "S", 1, "s", 1);
    MapData * progNameNode     = addMapMembers(&paramMap_loc, "default" , STR , true , "Ssd"  , STRVIEW("prog-name"         ), "p", 1);
    setNodeNegation(maintainArgvNode, overrideArgvNode);
    setNodeNegation(overrideArgvNode, maintainArgvNode);
    setNodeNegation(useNamespaceNode,  noNamespaceNode);
    setNodeNegation( noNamespaceNode, useNamespaceNode);
    //helpMsgNode->data.required = false;     // this isnt really super required

    // TODO should clparser parse given input args if they don't match clparser args?
    //  - this would sorta be nice but would just be way to error prone that it wouldn't really be worth it
    Errors retVal = parseArgs(argc - 1, argv + 1, &flagMap_loc, &paramMap_loc, NULL, &passedArgs, false);
    if(retVal != Success){
        fprintf(stderr, "error parsing args: %d\n", retVal);
        return retVal;
    }

    // set the default value for the shell if not given
    const char * shellStr =    shellNode->data.ptr;
    const char * progName = progNameNode->data.ptr;
    char parentCommandStr[MAX_SHELL_LEN] = "";
    bool allocatedProgName = false;
    pid_t ppid = getppid();
    if(strcmp(shellStr, "default") == 0){
        // TODO figure out a general way of getting the calling process name
        char pcmdFilename[MAX_PROCCOMM_LEN];
        snprintf(pcmdFilename, MAX_PROCCOMM_LEN, "/proc/%d/comm", ppid);
        FILE * pcmdFile = fopen(pcmdFilename, "r");
        fgets(parentCommandStr, MAX_SHELL_LEN, pcmdFile);
        int comlen = strlen(parentCommandStr);
        if( parentCommandStr[comlen-1] == '\n'){
            parentCommandStr[comlen-1] =  '\000';
        }
        // TODO xonsh seems to be more of a pain (appimage super doesnt work like this, actual install puts part of the script name in comm?)
        shellStr = parentCommandStr;
        shellNode->data.defaultData->ptr = shellStr;
    }
    if(strcmp(progName, "default") == 0){
        // TODO figure out a general way of getting the calling script
        char pcmdFilename[MAX_PROCCMDLINE_LEN];
        snprintf(pcmdFilename, MAX_PROCCMDLINE_LEN, "/proc/%d/cmdline", ppid);
        FILE * pcmdFile = fopen(pcmdFilename, "r");
        size_t n = 0;
        char * cmdline = NULL;
        // call getline twice to get the first arg (script filename)
        getdelim(&cmdline, &n, '\000', pcmdFile);
        getdelim(&cmdline, &n, '\000', pcmdFile);
        progName = cmdline;
        progNameNode->data.defaultData->ptr = progName;
        allocatedProgName = progName != NULL;
    }

    if(*(const bool *)helpNode->data.ptr){
        // TODO have the help info not print out to stderr
        printUsage(&flagMap_loc, &paramMap_loc, NULL, argv[0]);
        fprintf(stderr, "\tclparser takes in command line arguments to parse as command line arguments and a specification\n");
        fprintf(stderr, "\tfor which command line arguments should exist from stdin, and outputs shell code to set variables\n");
        fprintf(stderr, "\tset from the command line arguments to stdout.  Command line arguments to be parsed should be\n");
        fprintf(stderr, "\tgiven after --.  This will prevent the command line arguments to be parsed from being interpreted\n");
        fprintf(stderr, "\tas clparser command line arguments.  See the man page or README.md for more details\n");
        printHelp(&flagMap_loc, &paramMap_loc, NULL,
                "help               = print this help message\n\
                 maintain-argv      = do not override argv\n\
                 override-argv      = do     override argv\n\
                 namespace          = namespace args by using associative arrays if available or prepending flags_ or params_ to the variable names\n\
                 no-namespace       = do not use associative arrays or prefixes on variable names\n\
                 no-output-empty    = don't output parameters that aren't given\n\
                 unknown-positional = if an unknown argument is seen, treat all following arguments as positional arguments\n\
                 shell              = which shell syntax to use (sh, bash, zsh, ksh, csh, fish, xonsh) (default to calling shell)\n\
                 prog-name          = name of program to display in --help usage message\n\
                 help-exits         = exit after printing the help message (default to path of script called (first arg of cmdline of calling shell))\n\
                 help-msg           = some sort of string of help messages for args");
        return 1;
    }

    for(passedArgc = 0; passedArgs[passedArgc] != NULL && passedArgc <= argc; ++passedArgc);

    const char * helpMessage = helpMsgNode->data.ptr;

    Shell shell;
    if      (strcmp(shellStr,    "sh") == 0 || strcmp(shellStr, "dash") == 0 || strcmp(shellStr, "ash") == 0){
        shell = SH;
    }else if(strcmp(shellStr,  "bash") == 0){
        shell = BASH;
    }else if(strcmp(shellStr,   "zsh") == 0){
        shell = ZSH;
    }else if(strcmp(shellStr,   "ksh") == 0){
        shell = KSH;
    }else if(strcmp(shellStr,  "tcsh") == 0 || strcmp(shellStr, "csh") == 0){
        shell = CSH;
    }else if(strcmp(shellStr,  "fish") == 0){
        shell = FISH;
    }else if(strcmp(shellStr, "xonsh") == 0){
        shell = XONSH;
    }

    bool helpExits =  *(const bool *)helpExitsNode   ->data.ptr;
    bool useArgv   =  *(const bool *)overrideArgvNode->data.ptr;
    if( !useArgv &&  !*(const bool *)maintainArgvNode->data.ptr && shell == SH){
        // if didnt set override or maintain argv, then default to maintain unless using SH since it has no arrays
        // (so the only way to maintain whitespace in sh is to use argv)
        useArgv = true;
    }
    bool useNamespace =   *(const bool *)useNamespaceNode->data.ptr;
    if( !useNamespace && !*(const bool *) noNamespaceNode->data.ptr){
        // if didnt set use/don't use namespace, then default to use namespace unless using SH, CSH, or FISH since they have no associative arrays
        switch(shell){
            case CSH:
            case FISH:
            case SH:
                useNamespace = false;
                break;
            case BASH:
            case ZSH:
            case KSH:
            case XONSH:
            default:
                useNamespace = true;
                break;
        }
    }
    ParsePrintOptions parseOpts = {
        .shell=shell,
        .useArgv=useArgv,
        .useNamespace=useNamespace,
        .noOutputEmpty=*(const bool *)noOutEmptyNode->data.ptr,
        .unknownPositional=*(const bool *)unkIsPosNode->data.ptr,
    };

    // free the maps, should be done with the info now
    freeMap(&flagMap_loc);
    freeMap(&paramMap_loc);

    char * cbuf = NULL;
    size_t n = 0;
    ssize_t len;
    // TODO should i just use getdelim with \0?
    len = getline(&cbuf, &n, stdin);
    // TODO check that there wasnt an error
    if(len == -1){
        perror("read spec");
        return 2;
    }
    map_t flagMap, paramMap, posMap;
    initMap(&flagMap);
    initMap(&paramMap);
    initMap(&posMap);

    // let me think about what im doing here
    // ferr is a pointer to the char after the ; ending the flags or whatever error happened
    char * ce = cbuf + len;
    // find all flag params, make the default value "0" (false)
    char * ferr = parseArgSpec(cbuf, &flagMap, "flags:", (void *)&flagFalse, BOOL, false, NULL);
    //printMap(&flagMap);

    //addMapMembers(&flagMap, (void *)&flagFalse, BOOL, "sdsd", "help", 4, "h", 1);

    // if the ferr is before or after the buff then we know theres no flags
    bool noflag = false;
    bool noparam = false;
    if(ferr == NULL){
        fprintf(stderr, "error reading flags\n");
        return 3;
    } else if(ferr <= cbuf || ferr > ce){
        noflag = true;
    }

    // puts("params now");
    char * perr = parseArgSpec(cbuf, &paramMap, "parameters:", NULL, STR, true, NULL);
    //printMap(&paramMap);
    // if didnt find params and no flag then return 1 this is bad
    if(perr == NULL){
        // if params errored on parsing then exit
        fprintf(stderr, "error reading parameters\n");
        return 3;
    } else if(perr <= cbuf || perr > ce){
        noparam = true;
    }

    MapData * * positionalParamNodes = NULL;
    perr = parseArgSpec(cbuf, &posMap, "positionals:", NULL, STR, true, &positionalParamNodes);
    //printMap(&paramMap);
    // if didnt find params and no flag then return 1 this is bad
    if(perr == NULL){
        // if params errored on parsing then exit
        fprintf(stderr, "error reading positionals\n");
        return 3;
    } else if(perr <= cbuf || perr > ce){
        // if didnt get pos params, then free it rather than having explicit 0 positionals defined
        free(positionalParamNodes);
        positionalParamNodes = NULL;
        if(noflag && noparam){
            fprintf(stderr, "found neither flags nor parameters nor positionals\n");
            return 2;
        }
    }

    retVal = parseArgsPrint(passedArgc, passedArgs, &flagMap, &paramMap, positionalParamNodes, &parseOpts);
    // TODO do some stuffs and figure out the errors
    //  - probably exit with this exit code specifically
    //if(retVal != Success){
    //    fprintf(stderr, "error parsing args: %d\n", retVal);
    //}

    //bool help = getMapMember_bool(&flagMap, "help", 4);
    helpNode = getMapNode(&flagMap, "help", 4);
    if(helpNode != NULL && *(const bool *)helpNode->data.ptr){
    //if(hasNode(&flagMap, "help", 4) && getMapMember_bool(&flagMap, "help", 4)){
        printUsage(&flagMap, &paramMap, positionalParamNodes, progName);
        if(helpMessage != NULL){
            printHelp(&flagMap, &paramMap, positionalParamNodes, helpMessage);
        }
        if(helpExits){
            printExit(shell);
        }
        return 0;
    }

    freeMap(&flagMap);
    freeMap(&paramMap);
    freeMap(&posMap);
    if(positionalParamNodes != NULL){
        free(positionalParamNodes);
    }

    free(cbuf);
    if(allocatedProgName){
        free((void *)progName);
    }

    return retVal;
}




